actors/builtin/market/deal.go

```diff
@@ -12,6 +24,167 @@ import (
 //}
 var PieceCIDPrefix = market0.PieceCIDPrefix
 
+// The DealLabel is a kinded union of string or byte slice.
+// It serializes to a CBOR string or CBOR byte string depending on which form it takes.
+// The empty label is serialized as an empty CBOR string (maj type 3).
+type DealLabel struct {
+	bs        []byte
+	notString bool
+}
+
+// Zero value of DealLabel is canonical EmptyDealLabel but three different values serialize the same way:
+// DealLabel{nil, nil}, (*DealLabel)(nil), and DealLabel{&"", nil}
+var EmptyDealLabel = DealLabel{}
+
+func NewLabelFromString(s string) (DealLabel, error) {
+	if len(s) > DealMaxLabelSize {
+		return EmptyDealLabel, xerrors.Errorf("provided string is too large to be a label (%d), max length (%d)", len(s), DealMaxLabelSize)
+	}
+	if !utf8.ValidString(s) {
+		return EmptyDealLabel, xerrors.Errorf("provided string is invalid utf8")
+	}
+	return DealLabel{
+		bs:        []byte(s),
+		notString: false,
+	}, nil
+}
+
+func NewLabelFromBytes(b []byte) (DealLabel, error) {
+	if len(b) > DealMaxLabelSize {
+		return EmptyDealLabel, xerrors.Errorf("provided bytes are too large to be a label (%d), max length (%d)", len(b), DealMaxLabelSize)
+	}
+
+	return DealLabel{
+		bs:        b,
+		notString: true,
+	}, nil
+}
+
+func (label DealLabel) IsString() bool {
+	return !label.notString
+}
+
+func (label DealLabel) IsBytes() bool {
+	return label.notString
+}
+
+func (label DealLabel) ToString() (string, error) {
+	if !label.IsString() {
+		return "", xerrors.Errorf("label is not string")
+	}
+
+	return string(label.bs), nil
+}
+
+func (label DealLabel) ToBytes() ([]byte, error) {
+	if !label.IsBytes() {
+		return nil, xerrors.Errorf("label is not bytes")
+	}
+	return label.bs, nil
+}
+
+func (label DealLabel) Length() int {
+	return len(label.bs)
+}
+
+func (l DealLabel) Equals(o DealLabel) bool {
+	return bytes.Equal(l.bs, o.bs) && l.notString == o.notString
+}
+
+func (label *DealLabel) MarshalCBOR(w io.Writer) error {
+	scratch := make([]byte, 9)
+
+	// nil *DealLabel counts as EmptyLabel
+	// on chain structures should never have a pointer to a DealLabel but the case is included for completeness
+	if label == nil {
+		if err := cbg.WriteMajorTypeHeaderBuf(scratch, w, cbg.MajTextString, 0); err != nil {
+			return err
+		}
+		if _, err := io.WriteString(w, string("")); err != nil {
+			return err
+		}
+		return nil
+	} else if label.IsString() && label.IsBytes() {
+		return fmt.Errorf("dealLabel cannot have both string and bytes set")
+	} else if label.IsBytes() {
+		if len(label.bs) > cbg.ByteArrayMaxLen {
+			return xerrors.Errorf("labelBytes is too long to marshal (%d), max allowed (%d)", len(label.bs), cbg.ByteArrayMaxLen)
+		}
+
+		if err := cbg.WriteMajorTypeHeaderBuf(scratch, w, cbg.MajByteString, uint64(len(label.bs))); err != nil {
+			return err
+		}
+
+		if _, err := w.Write((label.bs)[:]); err != nil {
+			return err
+		}
+	} else { // label.IsString()
+		if len(label.bs) > cbg.MaxLength {
```

Comment:
Se use `MaxByteArrayLen` in both cases (on further inspection). I'd just lift this case to the top.